sdk/eventhub/azure-eventhubs/azure/eventhub/_client_base.py

```diff
@@ -0,0 +1,376 @@
+            output['path'] = eh_info[b'name'].decode('utf-8')
+            output['created_at'] = datetime.datetime.utcfromtimestamp(float(eh_info[b'created_at']) / 1000)
+            output['partition_ids'] = [p.decode('utf-8') for p in eh_info[b'partition_ids']]
+        return output
+
+    def get_partition_ids(self):
+        # type:() -> List[str]
+        """
+        Get partition ids of the specified EventHub.
+
+        :rtype: list[str]
+        :raises: :class:`EventHubError<azure.eventhub.EventHubError>`
+        """
+        return self.get_properties()['partition_ids']
+
+    def get_partition_properties(self, partition):
+        # type:(str) -> Dict[str, Any]
+        """
+        Get properties of the specified partition async.
+        Keys in the details dictionary include:
+
+            - event_hub_path
+            - id
+            - beginning_sequence_number
+            - last_enqueued_sequence_number
+            - last_enqueued_offset
+            - last_enqueued_time_utc
+            - is_empty
+
+        :param partition: The target partition id.
+        :type partition: str
+        :rtype: dict
+        :raises: :class:`EventHubError<azure.eventhub.EventHubError>`
+        """
+        mgmt_msg = Message(application_properties={'name': self.eh_name,
+                                                   'partition': partition})
+        response = self._management_request(mgmt_msg, op_type=MGMT_PARTITION_OPERATION)
+        partition_info = response.get_data()
+        output = {}
+        if partition_info:
+            output['event_hub_path'] = partition_info[b'name'].decode('utf-8')
+            output['id'] = partition_info[b'partition'].decode('utf-8')
+            output['beginning_sequence_number'] = partition_info[b'begin_sequence_number']
+            output['last_enqueued_sequence_number'] = partition_info[b'last_enqueued_sequence_number']
+            output['last_enqueued_offset'] = partition_info[b'last_enqueued_offset'].decode('utf-8')
+            output['last_enqueued_time_utc'] = datetime.datetime.utcfromtimestamp(
+                float(partition_info[b'last_enqueued_time_utc'] / 1000))
+            output['is_empty'] = partition_info[b'is_partition_empty']
+        return output
+
+    def close(self):
+        # type:() -> None
+        self._conn_manager.close_connection()
+
+
+class ConsumerProducerMixin(object):
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def _check_closed(self):
+        if self.closed:
+            raise EventHubError("{} has been closed. Please create a new one to handle event data.".format(self._name))
+
+    def _open(self):
+        """Open the EventHubConsumer/EventHubProducer using the supplied connection.
+
+        """
+        # pylint: disable=protected-access
+        if not self.running:
+            if self._handler:
+                self._handler.close()
```

Comment:
The encapsulation is broken/mismatched here. On one hand, we are directly looking at the `self._handler` attribute. On the other hand, we are calling a method `_create_handler()` that sets the `self._handler` attribute.

Reply:
Also, you have a `_close_handler`, so if finally you actually close it, call your own method and don't duplicate code.